apps/server/lib/lexical/server/code_intelligence/entity.ex

```diff
@@ -50,55 +53,141 @@ defmodule Lexical.Server.CodeIntelligence.Entity do
-    module_string =
-      charlist
-      |> Enum.with_index(column)
-      |> Enum.take_while(fn {char, column} ->
-        column < position.character or char != ?.
-      end)
-      |> Enum.map(&elem(&1, 0))
-      |> List.to_string()
-
-    expanded =
-      [module_string]
-      |> Module.concat()
-      |> Ast.expand_aliases(document, position)
-
-    with {:ok, module} <- expanded do
-      {:ok, {:module, module}, {{line, column}, {line, column + String.length(module_string)}}}
+  defp resolve_module(charlist, {{line, column}, {line, _}}, document, position) do
+    module_string = module_before_position(charlist, column, position)
+
+    with {:ok, module} <- expand_alias(module_string, document, position) do
+      end_column = column + String.length(module_string)
+      {:ok, {:module, module}, {{line, column}, {line, end_column}}}
     end
   end
 
   # Modules on multiple lines, e.g. "Foo.\n  Bar.\n  Baz"
   # Since we no longer have formatting information at this point, we
   # just return the entire module for now.
   defp resolve_module(charlist, node_range, document, position) do
-    module_string = List.to_string(charlist)
+    with {:ok, module} <- expand_alias(charlist, document, position) do
+      {:ok, {:module, module}, node_range}
+    end
+  end
+
+  # Take only the segments at and before the cursor, e.g.
+  # Foo|.Bar.Baz -> Foo
+  # Foo.|Bar.Baz -> Foo.Bar
+  defp module_before_position(charlist, start_column, position) when is_list(charlist) do
+    charlist
+    |> List.to_string()
+    |> module_before_position(position.character - start_column)
+  end
 
-    expanded =
-      [module_string]
-      |> Module.concat()
-      |> Ast.expand_aliases(document, position)
+  defp module_before_position(string, index) when is_binary(string) do
+    {prefix, suffix} = String.split_at(string, index)
 
-    with {:ok, module} <- expanded do
-      {:ok, {:module, module}, node_range}
+    case String.split(suffix, ".", parts: 2) do
+      [before_dot, _after_dot] -> prefix <> before_dot
+      [before_dot] -> prefix <> before_dot
+    end
+  end
+
+  defp expand_alias({:alias, {:local_or_var, prefix}, charlist}, document, position) do
+    expand_alias(prefix ++ [?.] ++ charlist, document, position)
+  end
+
+  defp expand_alias({:alias, charlist}, document, position) do
+    expand_alias(charlist, document, position)
+  end
+
+  defp expand_alias(charlist, document, position) when is_list(charlist) do
+    charlist
+    |> List.to_string()
+    |> expand_alias(document, position)
+  end
+
+  defp expand_alias(module, document, position) when is_binary(module) do
+    [module]
+    |> Module.concat()
+    |> Ast.expand_aliases(document, position)
+  end
+
+  defp expand_alias(_, _document, _position), do: :error
+
+  # Pipes:
+  defp arity_at_position([{:|>, _, _} = pipe | _], position) do
+    {_call, _, args} =
+      pipe
+      |> Macro.unpipe()
+      |> Stream.map(fn {ast, _arg_position} -> ast end)
```

Comment:
Yes, this prevents an additional list from being created, since this returns a lazy enumerable that will only yield a mapped value when "requested" during the `Enum.find`. 

That said, I should actually be using [`Enum.find_value`](https://hexdocs.pm/elixir/1.15.5/Enum.html#find_value/3) here instead.